File: Models/Channel.cs
using B4mServer.Websockets;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace B4mServer.Models;

public partial class Channel
{
	public int Id { get; set; }
	public string Name { get; set; } = null!;
	public long Created { get; set; }
	public string Color { get; set; } = null!;
	[JsonIgnore]
	public string? Password { get; set; }

	// foreign key
	public int OwnerId { get; set; }

	// Navigational Properties
	[JsonIgnore]
	[ForeignKey("OwnerId")]
	public virtual User? Owner { get; set; }
	[JsonIgnore]
	public virtual ICollection<Message> Messages { get; set; } = [];

	public WebSocketChannel ToDTO()
	{
		return new WebSocketChannel
		{
			Id = Id,
			Name = Name,
			Created = Created,
			Color = Color,
			OwnerId = OwnerId,
			Messages = Messages.Select(m => m.ToMinimalDTO()).ToList()
		};
	}
}

File: frontend/src/types/channel.ts
import { User } from './user';
import { Message } from './message';

export class Channel {
	id: number;
	name: string;
	users: User[] = [];
	messages: Message[] = [];
	password?: string;
	created: number;
	color: string;
	ownerId?: number;

	constructor(
		id: number,
		name: string,
		color: string,
		created: number = Date.now(),
		ownerId?: number,
		password?: string
	) {
		this.id = id;
		this.name = name;
		this.created = created;
		this.color = color;
		this.id = id;
		this.ownerId = ownerId;
		this.password = password;
	}

	getSerialized() {
		return {
			id: this.id,
			name: this.name,
			users: this.users.map(user => user.getSerialized()),
			messages: this.messages.map(message => message.getSerialized()),
			password: this.password,
			created: this.created,
			color: this.color,
			ownerId: this.ownerId
		};
	}
	
	toDTO() {
		return new Channel(this.id, this.name, this.color, this.created, this.ownerId, this.password);
	}
}

export class MinChannel {
	id: number;
	name: string;
	created: number;
	color: string;
	ownerId?: number;

	constructor(
		id: number,
		name: string,
		color: string,
		created: number = Date.now(),
		ownerId?: number,
	) {
		this.id = id;
		this.name = name;
		this.created = created;
		this.color = color;
		this.id = id;
		this.ownerId = ownerId;
	}
}

export class NewChannel {
	name: string = 'untitled';
	color: string = 'grey';
	password?: string;
}